Make ScriptSortBuilder implement Writable

cbuescher opened the pull request:
This adds methods and tests to ScriptSortBuilder that makes it implement NamedWritable and adds the fromXContent method needed to read itseld from xContent.

cbuescher committed:
Make ScriptSortBuilder implement NamedWritable

This adds methods and tests to ScriptSortBuilder that
makes it implement NamedWritable and adds the fromXContent
method needed to read itseld from xContent.

cbuescher commented:
This also relates to #15178, mind to take a look @MaineC ?

cbuescher committed:
Added enum for script sort type

MaineC commented on core/src/test/java/org/elasticsearch/search/sort/ScriptSortBuilderTests.java:
```diff
@@ -0,0 +1,104 @@
+        assertEquals(ScriptSortType.STRING, ScriptSortType.fromString("STRING"));
+        assertEquals(ScriptSortType.NUMBER, ScriptSortType.fromString("number"));
+        assertEquals(ScriptSortType.NUMBER, ScriptSortType.fromString("Number"));
+        assertEquals(ScriptSortType.NUMBER, ScriptSortType.fromString("NUMBER"));
```
Maybe add a test or two that check the exceptions you are throwing in case of invalid values? Otherwise the string to enum part of the PR looks good.

cbuescher replied:
done

cbuescher committed:
Merge branch 'master' into sort-serialization-scriptsort

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/ScriptSortBuilder.java:
```diff
@@ -78,22 +146,189 @@ public ScriptSortBuilder setNestedPath(String nestedPath) {
+    /**
+     * Gets the nested path.
+     */
+    public String getNestedPath() {
```
Nitpick: This mixes plain old POJO getters/setters with fluent interface like getters/setters. In another PR I left them untouched as well. Maybe at least track in a separate issue to unify those?

cbuescher replied:
Probably this issue: https://github.com/elastic/elasticsearch/issues/14266.

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/ScriptSortBuilder.java:
```diff
@@ -78,22 +146,189 @@ public ScriptSortBuilder setNestedPath(String nestedPath) {
-        builder.startObject("_script");
-        builder.field("script", script);
-        builder.field("type", type);
+        builder.startObject(NAME);
```
Great seeing these replaced with ParseFields.

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/ScriptSortBuilder.java:
```diff
@@ -78,22 +146,189 @@ public ScriptSortBuilder setNestedPath(String nestedPath) {
+                    sortMode = parser.text();
+                } else if (parseField.match(currentName, NESTED_PATH_FIELD)) {
+                    nestedPath = parser.text();
+                }
```
Exception for fall through? (Thought about it after reading the comment in line 205)

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/ScriptSortBuilder.java:
```diff
@@ -78,22 +146,189 @@ public ScriptSortBuilder setNestedPath(String nestedPath) {
+                    params = parser.map();
+                } else if (parseField.match(currentName, NESTED_FILTER_FIELD)) {
+                    nestedFilter = context.parseInnerQueryBuilder();
+                }
```
Exception for fall through?

MaineC commented:
Left a few minor comments but overall looks good.

cbuescher committed:
Adding enum for SortMode and use it in ScriptSortBuilder

cbuescher committed:
Adressing review comments, adding parsing tests

cbuescher commented:
@MaineC thanks, I added a commit that switches sort mode to an enum and addressed your review comments. Let me know if you think there is anything to add.

MaineC commented:
LGTM - nothing to add from my side (apart from rebasing)

cbuescher committed:
Merge branch 'master' into sort-serialization-scriptsort

 Conflicts:
	core/src/test/java/org/elasticsearch/search/sort/AbstractSortTestCase.java

cbuescher committed:
Using SortMode enum in all sort builders

cbuescher commented:
@MaineC thanks, after adding the SortMode enum in ScriptSortBuilder I also switched GeoDistanceSortBuilder and FieldSortBuilder to using it. Maybe you can take a last look at those changes.
